### PR TITLE
🚀 Feature: Ability to report a message in connect.

### PR DESCRIPTION
**Pull request by sarthology**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
The fastest implementation of this feature request that I can think of:
- Now, one can report a message from others.
- And Block them also after reporting.
- Custom Report for admin to see the connect feedbacks.

## QA Instructions, Screenshots, Recordings
Screen Recording 2020-11-09 at 11 35 AM

## Added tests?

- [ ] Yes
- [x] No, and this is why: _Will add soon_
- [ ] I need help with writing tests

## Added to documentation?

- [ ] Docs.forem.com
- [ ] README
- [x] No documentation needed

##  What gif best describes this PR or how it makes you feel?

alt_text

### CODE REVIEW

File: app/controllers/feedback_messages_controller.rb [hunk 1]
```diff
@@ -19,12 +18,24 @@ def create
...
-
       flash[:notice] = "Make sure the forms are filled 🤖"
-      render "pages/report_abuse"
+
+      respond_to do |format|
+        format.html { render "pages/report_abuse" }
+        format.json do
+          render json: {
+            success: false,
+            message: @feedback_message.errors.full_messages.to_sentence,
```
Comment (citizen428) [inline app/controllers/feedback_messages_controller.rb:34]: You can do `@feedback_message.errors_as_sentence` here, we defined that on `ApplicationRecord`.

**Comment by citizen428** [comment]
I'll play around with this later/tomorrow, but thanks for the quick turnaround time @sarthology! 🙏

File: app/javascript/chat/ReportAbuse/index.jsx [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+  const handleSubmit = async () => {
+    const response = await reportAbuse(
+      data.message,
+      'connect',
+      category,
+      data.user_id,
+    );
+    const { success, message } = response;
+    if (success) {
+      const confirmBlock = window.confirm(
```
Comment (nickytonline) [inline app/javascript/chat/ReportAbuse/index.jsx:24]: Refactor (non-blocking): Use the [Modal](https://storybook.forem.com/?path=/story/3-components-modals--default) component instead. You can use a Button Group component for the OK/Cancel and you can await a boolean value.

If this is too much work for now, we can do this in another PR. In fact we should probably have  additional props on the component for action buttons, cc: @ludwiczakpawel

**Review by nickytonline — CHANGES_REQUESTED** [review]
Starting to shape up. Some request changes and also, the PR will need some tests for the frontend and potentially the backend (not positive).

**Comment by citizen428** [comment]
Hm, while I like the idea of being able to block the user who wrote the message I just reported, the current implementation still has some rough edges. When I click "cancel" (I only want to report but not block), it doesn't _seem_ like the form gets submitted. However, when trying to submit again after hitting "cancel" on the popup it shows a validation error "Reporter has already been taken". I confirmed via `/admin/reports` that the report hs been created.

File: app/controllers/feedback_messages_controller.rb [hunk 3]
```diff
@@ -8,8 +8,7 @@ def create
 
     params = feedback_message_params.merge(reporter_id: current_user&.id)
     @feedback_message = FeedbackMessage.new(params)
-
-    if (recaptcha_disabled? || recaptcha_verified?) && @feedback_message.save
+    if (recaptcha_disabled? || recaptcha_verified? || params[:feedback_type] == "connect") && @feedback_message.save
```
Comment (citizen428) [inline app/controllers/feedback_messages_controller.rb:11]: Maybe you want to wrap up `params[:feedback_type] == "connect"` in a new predicate method too?

```suggestion
    if (recaptcha_disabled? || recaptcha_verified? || connect_feedback?) && @feedback_message.save
```

**Comment by citizen428** [comment]
For tests, I think I'd like to see at least one system spec with a change assertion that verifies that a feedback message was created. Something like:

```ruby
expect do
  visit ...
   # capaybara stuff goes here
end.to change(FeedbackMessage, :count).by(1)

File: app/javascript/chat/ReportAbuse/index.jsx (hunk 2 above)
Comment (ludwiczakpawel) [reply app/javascript/chat/ReportAbuse/index.jsx:24]: Yea. I'd definitely postpone it. We've been working on some design updates for Connect so I'd rather ship stuff looking not perfect right now and revisit the UI a bit later, definitely in a separate PR.

**Comment by narender2031** [comment]
> Hm, while I like the idea of being able to block the user who wrote the message I just reported, the current implementation still has some rough edges. When I click "cancel" (I only want to report but not block), it doesn't _seem_ like the form gets submitted.

I fixed rough edges for the PR.  
> When I click "cancel" (I only want to report but not block), it doesn't _seem_ like the form gets submitted. 

This is also handled.

>  However, when trying to submit again after hitting "cancel" on the popup it shows a validation error "Reporter has already been taken".

User can only create One report.

**Comment by citizen428** [comment]
@narender2031 

> User can only create One report.

Yes, that's good. What I meant was that there was no indication previously that a report was actually created after hitting cancel on the popup, so re-submitting seemed like something a lot of people would try.

The current implementation is better, but I think the message in the bottom left corner is somewhat easy to miss. Can it be placed elsewhere? Also it may make sense to close the sidebar after a successful submission, it serves no further purpose so having the user perform an extra action seems a bit superfluous.

**Comment by narender2031** [comment]
> @narender2031
> 
> > User can only create One report.
> 
> Yes, that's good. What I meant was that there was no indication previously that a report was actually created after hitting cancel on the popup, so re-submitting seemed like something a lot of people would try.
> 
> The current implementation is better, but I think the message in the bottom left corner is somewhat easy to miss. Can it be placed elsewhere? Also it may make sense to close the sidebar after a successful submission, it serves no further purpose so having the user perform an extra action seems a bit superfluous.

@citizen428  I was also thinking the same but, In an `addSnackbarItem` we cannot change the position for message. 
Another possible solution is to create a custom message UI close to the sidebar. Please add your suggestion.

**Comment by citizen428** [comment]
Ah, ok, Snackbar is fine then, maybe we can automatically close the sidebar at least?

**Comment by narender2031** [comment]
> Ah, ok, Snackbar is fine then, maybe we can automatically close the sidebar at least?

Yes. I think we can do this. I will do this today .

**Comment by narender2031** [comment]
@citizen428 PR is updated

File: app/javascript/chat/ReportAbuse/index.jsx [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+        <h1 className="lh-tight mb-4 mt-0">Report Abuse</h1>
+        <p>
+          Thank you for reporting any abuse that violates our{' '}
+          <a href="/code-of-conduct">code of conduct</a> or
+          <a href="/terms">terms and conditions</a>. We continue to try to make
+          environment a great one for everybody.
+        </p>
+      </div>
+      <div className="crayons-card crayons-card--secondary p-4">
+        <div className="crayons-fields">
```
Comment (nickytonline) [inline app/javascript/chat/ReportAbuse/index.jsx:58]: refactor (blocking): use a semantic element here, most likely [fieldset](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/fieldset). Any other suggestions in this component @marcysutton to alleviate `<div />` soup?

File: app/javascript/chat/ReportAbuse/index.jsx [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+        addSnackbarItem({ message: 'Your report has been submitted.' });
+      }
+      closeReportAbuseForm();
+    } else {
+      addSnackbarItem({ message });
+    }
+  };
+
+  return (
+    <div className="p4">
```
Comment (nickytonline) [inline app/javascript/chat/ReportAbuse/index.jsx:47]: refactor (potentially blocking): There is heavy use of `<div />`'s in this component. One, are all of these wrapper `<div />`'s necessary? cc: @ludwiczakpawel

File: app/javascript/chat/ReportAbuse/index.jsx [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+      <div className="p-4 grid gap-2 crayons-card mb-4">
```
Comment (nickytonline) [inline app/javascript/chat/ReportAbuse/index.jsx:48]: refactor (blocking): These sections of content should use the [section](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/section) element instead.

**Review by nickytonline — CHANGES_REQUESTED** [review]
I have some more request changes @narender2031.

**Comment by narender2031** [comment]
> I have some more request changes @narender2031.

all requested changes are done

**Review by nickytonline — CHANGES_REQUESTED** [review]

File: app/javascript/chat/ReportAbuse/index.jsx [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+import { reportAbuse, blockUser } from '../actions/requestActions';
+import { addSnackbarItem } from '../../Snackbar';
+import { Button, FormField, RadioButton } from '@crayons';
+
+/**
+ * This component render the report abuse
+ *
+ * @param {object} props
+ * @param {object} props.data
+ * @param {object} props.closeReportAbuseForm
```
[the remark this answers is not in the document]
Comment (nickytonline) [reply app/javascript/chat/ReportAbuse/index.jsx:13]: I'd make the change myself, but for some reason, I can't apply the suggested change.

**Comment by nickytonline** [comment]
I'm having issues on my local with the branch. I can't seem to get the feed to load. Also, we've moved to node 14.14 and Ruby 2.7.2. Can you merge latest of master into your branch @narender2031? I'd push the change myself, but I do not seem to be able to push changes to your branch. Thanks!

File: app/javascript/chat/ReportAbuse/index.jsx [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+        <h1 className="lh-tight mb-4 mt-0">Report Abuse</h1>
+        <p>
+          Thank you for reporting any abuse that violates our{' '}
+          <a href="/code-of-conduct">code of conduct</a> or
+          <a href="/terms">terms and conditions</a>. We continue to try to make
+          environment a great one for everybody.
+        </p>
+      </section>
+      <section className="crayons-card crayons-card--secondary p-4 justify-between">
+        <FormField variant="radio">
```
Comment (marcysutton) [inline app/javascript/chat/ReportAbuse/index.jsx:75]: Do these radio buttons get rendered in a `fieldset`? That would help to group them.

File: app/javascript/chat/message.jsx [hunk 9]
```diff
@@ -60,6 +61,22 @@ const Message = ({
       </div>
     </div>
   );
+  const dropdownReport = (
+    <div className="message__actions">
+      <span className="ellipsis__menubutton">
+        <img src={ThreeDotsIcon} alt="dropdown menu icon" />
```
Comment (marcysutton) [inline app/javascript/chat/message.jsx:67]: I can't seem to set my local instance up to test this properly, but from looking at the markup it appears this isn't an accessible dropdown (`span` and `img` elements are not focusable or keyboard operable). How does this work from a keyboard/screen reader when rendered?

**Review by marcysutton — CHANGES_REQUESTED** [review]
Left some accessibility feedback since I was tagged by @nickytonline.

Are there tips on how to test this locally? Do you have to create multiple users somehow?

File: app/javascript/chat/ReportAbuse/index.jsx (hunk 8 above)
Comment (nickytonline) [reply app/javascript/chat/ReportAbuse/index.jsx:75]: Good catch @marcysutton, they do not render in a [fieldset](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/fieldset). @narender2031, can you make the change Marcy is suggesting?

**Review by nickytonline — CHANGES_REQUESTED** [review]

File: app/javascript/chat/ReportAbuse/index.jsx (hunk 8 above)
Comment (narender2031) [reply app/javascript/chat/ReportAbuse/index.jsx:75]: done

**Comment by citizen428** [comment]
@nickytonline @marcysutton Thanks for all the A11y feedback, much appreciated. Given we currently lack abuse reports in Connect, how do you feel about merging this now and then making whatever accessibility improvements we are still missing in a follow-up PR?

File: spec/requests/feedback_messages_spec.rb [hunk 10]
```diff
@@ -50,7 +50,7 @@ def mock_recaptcha_config_enabled
         end
       end
     end
-
+
```
Comment (citizen428) [inline spec/requests/feedback_messages_spec.rb:53]: **nitpick (blocking):** Please clean up this whitespace change.

File: spec/requests/feedback_messages_spec.rb [hunk 11]
```diff
@@ -68,8 +68,8 @@ def mock_recaptcha_config_enabled
         expect(FeedbackMessage.where(offender_id: user.id).count).to eq(1)
       end
     end
-
-    context "with no recaptcha keys set" do
+
```
Comment (citizen428) [inline spec/requests/feedback_messages_spec.rb:71]: **nitpick (blocking):** Please clean up this whitespace change.

**Review by citizen428 — CHANGES_REQUESTED** [review]

**Review by nickytonline — CHANGES_REQUESTED** [review]
Just  following up on a previous change request @narender2031.

**Comment by citizen428** [comment]
@nickytonline I think @narender2031 is still on leave.

**Comment by sarthology** [comment]
@citizen428 @nickytonline Something went wrong after merge from Master, fixing it currently.

File: spec/requests/feedback_messages_spec.rb (hunk 11 above)
Comment (narender2031) [reply spec/requests/feedback_messages_spec.rb]: @citizen428 These changes are done

**Comment by narender2031** [comment]
> > Good catch @marcysutton, they do not render in a [fieldset](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/fieldset). @narender2031, can you make the change Marcy is suggesting?
> 
> Just following up on a previous change request @narender2031.

@nickytonline  These changes are done. all the radio buttons are grouped under one `fieldset`.

**Comment by narender2031** [comment]
@citizen428 & @nickytonline Travis tests are failing due to `knapsack ` issue. can you please check

**Comment by nickytonline** [comment]
> @citizen428 & @nickytonline Travis tests are failing due to `knapsack ` issue. can you please check

Taking a peek @narender2031

**Comment by nickytonline** [comment]
I can't push to your branch @narender2031. Can you merge latest from master and push that up? We have Cypress in the project now and your branch is missing those changes, Thanks!

**Comment by narender2031** [comment]
> I can't push to your branch @narender2031. Can you merge latest from master and push that up? We have Cypress in the project now and your branch is missing those changes, Thanks!

Updated my branch with master but still tests are failing.

**Comment by citizen428** [comment]
@narender2031 This is temporarily blocked by a CI configuration issue, we need https://github.com/forem/forem/pull/11946 to go in first.

**Comment by rhymes** [comment]
@narender2031 please go ahead and merge master, https://github.com/forem/forem/pull/11946 is in

**Comment by nickytonline** [comment]
There appears to be a valid failure in the rspecs

```bash
Failures:

  1) Feedback report by chat channel messages when user create a report abuse feedback message feedback messahe should increase by one

     Failure/Error: params.require(:feedback_message).permit(allowed_params)

     

     ActionController::ParameterMissing:

       param is missing or the value is empty: feedback_message

     

     [Screenshot]: /home/travis/build/forem/forem/tmp/screenshots/failures_r_spec_example_groups_feedback_report_by_chat_channel_messages_when_user_create_a_report_abuse_feedback_message_feedback_messahe_should_increase_by_one_688.png

     

     # ./app/controllers/feedback_messages_controller.rb:57:in `feedback_message_params'

     # ./app/controllers/feedback_messages_controller.rb:9:in `create'

     # ./app/middlewares/set_time_zone.rb:10:in `call'

     # ./app/middlewares/set_cookie_domain.rb:13:in `call'

     # ./spec/system/feedback_message_spec.rb:13:in `block (4 levels) in <top (required)>'

     # ./spec/system/feedback_message_spec.rb:12:in `block (3 levels) in <top (required)>'

     # ./spec/rails_helper.rb:141:in `block (3 levels) in <top (required)>'

     # ./spec/rails_helper.rb:141:in `block (2 levels) in <top (required)>'

     # ./spec/support/initializers/ci_csv_formatter.rb:92:in `block (2 levels) in <top (required)>'

Finished in 5 minutes 48 seconds (files took 11.9 seconds to load)

1841 examples, 1 failure, 4 pending

Failed examples:

rspec ./spec/system/feedback_message_spec.rb:11 # Feedback report by chat channel messages when user create a report abuse feedback message feedback messahe should increase by one
```

**Comment by narender2031** [comment]
@nickytonline test case fixed

**Comment by nickytonline** [comment]
There is this failure as well @narender2031.

```bash
Failures:

  1) feedback_messages POST /feedback_messages when feedback is created by chat creates a feedback message

     Failure/Error: params.require(:feedback_message).permit(allowed_params)

     

     ActionController::ParameterMissing:

       param is missing or the value is empty: feedback_message

     # ./app/controllers/feedback_messages_controller.rb:57:in `feedback_message_params'

     # ./app/controllers/feedback_messages_controller.rb:9:in `create'

     # ./app/middlewares/set_time_zone.rb:10:in `call'

     # ./app/middlewares/set_cookie_domain.rb:13:in `call'

     # ./spec/requests/feedback_messages_spec.rb:57:in `block (4 levels) in <top (required)>'

     # ./spec/rails_helper.rb:141:in `block (3 levels) in <top (required)>'

     # ./spec/rails_helper.rb:141:in `block (2 levels) in <top (required)>'

Finished in 8 minutes 8 seconds (files took 7.8 seconds to load)

1651 examples, 1 failure, 2 pending

Failed examples:

rspec ./spec/requests/feedback_messages_spec.rb:65 # feedback_messages POST /feedback_messages when feedback is created by chat creates a feedback message
```

**Comment by narender2031** [comment]
@citizen428 all the test cases are passing now. I think we are good to go now.

**Comment by narender2031** [comment]
@citizen428 any updates on this PR?

**Comment by nickytonline** [comment]
I will give this another review @narender2031  now that those specs have been fixed.

**Comment by msarit** [comment]
Hey @sarthology! 👋🏾 

Just alerting you re: conflicts. Please resolve thanks!

**Comment by narender2031** [comment]
> Hey @sarthology! 👋🏾
> 
> Just alerting you re: conflicts. Please resolve thanks!

I'm fixing this in while.

**Comment by narender2031** [comment]
There is an unrelated test that is failing, can you please re-run the test cases
@citizen428 @nickytonline @msarit

**Comment by rhymes** [comment]
Re-ran

**Review by nickytonline — COMMENTED** [review]
@narender2031, can you allow me to push changes to your branch? I don't seem to have access to do it. Thanks.

**Comment by citizen428** [comment]
@nickytonline Worst case we can just close this PR and you open a new one from your local branch, WDYT?

**Comment by narender2031** [comment]
@nickytonline updated the rule for the branch now you can push the code for this branch

**Comment by nickytonline** [comment]
@citizen428, they gave me access so I'll finish up this work on this PR.

**Comment by nickytonline** [comment]
@citizen428, I'm going to close this as I'm still having issues pushing to the branch and will create a new one.
